File: android-tree-sitter/src/main/java/com/itsaky/androidide/treesitter/TSParser.java
/*
 *  This file is part of android-tree-sitter.
 *
 *  android-tree-sitter library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2.1 of the License, or (at your option) any later version.
 *
 *  android-tree-sitter library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *  along with android-tree-sitter.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.itsaky.androidide.treesitter;

import com.itsaky.androidide.treesitter.string.UTF16String;
import com.itsaky.androidide.treesitter.string.UTF16StringFactory;
import com.itsaky.androidide.treesitter.util.TSObjectFactoryProvider;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.locks.ReentrantLock;

/**
 * Implementation of tree sitter's <code>TSParser</code> APIs. This implementation always converts
 * the input source code to {@link UTF16String UTF-16 string}.
 */
public class TSParser extends TSNativeObject {

  protected final ReentrantLock parseLock = new ReentrantLock(true);
  protected final ReentrantLock cancellationLock = new ReentrantLock(true);
  protected final AtomicBoolean isParsing = new AtomicBoolean(false);
  protected final AtomicBoolean isCancellationRequested = new AtomicBoolean(false);

  protected TSParser(long pointer) {
    super(pointer);
  }

  protected TSParser() {
    this(Native.newParser());
  }

  public static TSParser create() {
    return create(Native.newParser());
  }

  public static TSParser create(long parserPointer) {
    return TSObjectFactoryProvider.getFactory().createParser(parserPointer);
  }

  private TSTree createTree(long pointer) {
    if (pointer == 0) {
      return null;
    }
    return TSTree.create(pointer);
  }

  /**
   * Set the language of the given parser.
   *
   * @param language The language to set.
   * @see TSLanguage
   */
  public void setLanguage(TSLanguage language) {
    checkAccess();
    Native.setLanguage(getNativeObject(), language.getNativeObject());
  }

  /**
   * Get the language for this parser instance.
   *
   * @return The language instance.
   */
  public TSLanguage getLanguage() {
    checkAccess();
    final var langPtr = Native.getLanguage(this.getNativeObject());
    if (langPtr == 0) {
      return null;
    }

    return TSLanguageCache.get(langPtr);
  }

  /**
   * Parses the given String source. See {@link #parseString(TSTree, UTF16String)} for more
   * details.
   *
   * @param source The source code to parse.
   * @return The parsed tree, or <code>null</code> if the parse failed or was cancelled.
   */
  public TSTree parseString(String source) {
    throwIfParseNotCancelled();
    try (final var str = UTF16StringFactory.newString(source)) {
      return parseString(str);
    }
  }

  /**
   * Parses the given {@link UTF16String} source. See {@link #parseString(TSTree, UTF16String)} for
   * more details.
   *
   * @param source The source code to parse.
   * @return The parsed tree, or <code>null</code> if the parse failed or was cancelled.
   */
  public TSTree parseString(UTF16String source) {
    return parseString(null, source);
  }

  /**
   * Parses the given source code bytes. See {@link #parseString(TSTree, UTF16String)} for more
   * details.
   *
   * @param bytes The source code to parse.
   * @return The parsed tree, or <code>null</code> if the parse failed or was cancelled.
   * @see #parseBytes(byte[], int, int)
   */
  public TSTree parseBytes(byte[] bytes) {
    return parseBytes(bytes, 0, bytes.length);
  }

  /**
   * Parses the given source code bytes. See {@link #parseString(TSTree, UTF16String)} for more
   * details. See {@link #parseString(TSTree, UTF16String)} for more details.
   *
   * @param bytes  The source code to parse.
   * @param offset The start offset in <code>bytes</code>.
   * @param len    The number of bytes to from <code>offset</code> to parse.
   * @return The parsed tree, or <code>null</code> if the parse failed or was cancelled.
   */
  public TSTree parseBytes(byte[] bytes, int offset, int len) {
    throwIfParseNotCancelled();
    try (final var source = UTF16StringFactory.newString(bytes, offset, len)) {
      return parseString(source);
    }
  }

  /**
   * Parse the given edited source code using the previously parsed syntax tree. The given
   * {@link TSTree} must have been edited using {@link TSTree#edit(TSInputEdit)} before calling this
   * method. See {@link #parseString(TSTree, UTF16String)} for more details.
   *
   * @param oldTree The previously parsed syntax tree.
   * @param source  The source code to parse.
   * @return The parsed tree, or <code>null</code> if the parse failed or was cancelled.
   */
  public TSTree parseString(TSTree oldTree, String source) {
    throwIfParseNotCancelled();
    try (final var str = UTF16StringFactory.newString(source)) {
      return parseString(oldTree, str);
    }
  }

  /**
   * Parse the given edited source code using the previously parsed syntax tree. The given
   * {@link TSTree} must have been edited using {@link TSTree#edit(TSInputEdit)} before calling this
   * method.
   * <p>
   * Throws {@link ParseInProgressException} if the parser is currently parsing a syntax tree and
   * the cancellation was NOT requested using {@link #requestCancellation()}. This method blocks the
   * current thread if the previous parse was requested to be cancelled but the parse operation has
   * not been cancelled yet.
   *
   * @param oldTree The previously parsed syntax tree.
   * @param source  The source code to parse.
   * @return The parsed tree, or <code>null</code> if the parse failed or was cancelled.
   * @throws IllegalStateException    If the parser is not accessible. See
   *                                  {@link TSNativeObject#canAccess()} for more details.
   * @throws ParseInProgressException If the parser is currently parsing another syntax tree.
   */
  public TSTree parseString(TSTree oldTree, UTF16String source) {
    checkAccess();

    // if the parser is currently parsing a syntax tree and the cancellation
    // was not requested, throw an error
    throwIfParseNotCancelled();

    // acquire the lock
    // this will wait until the cancelled parse call returns
    parseLock.lock();
    setCancellationRequested(false);
    setParsingFlag();
    try {
      final var strPointer = source.getPointer();
      final var oldTreePointer = oldTree != null ? oldTree.getNativeObject() : 0;
      final var tree = Native.parse(this.getNativeObject(), oldTreePointer, strPointer);
      return createTree(tree);
    } finally {
      unsetParsingFlag();
      parseLock.unlock();
    }
  }

  /**
   * Set the maximum duration in microseconds that parsing should be allowed to take before
   * halting.
   *
   * <p>If parsing takes longer than this, it will halt early, returning <code>null</code>.
   */
  public void setTimeout(long microseconds) {
    checkAccess();
    Native.setTimeout(getNativeObject(), microseconds);
  }

  /**
   * Get the duration in microseconds that parsing is allowed to take.
   *
   * @return The timeout in microseconds.
   */
  public long getTimeout() {
    checkAccess();
    return Native.getTimeout(getNativeObject());
  }

  /**
   * Check whether the parser is in the process of parsing a syntax tree.
   *
   * @return <code>true</code> if the parser is parsing a syntax tree, <code>false</code> otherwise.
   */
  public boolean isParsing() {
    return isParsing.get();
  }

  /**
   * Sets the 'parsing' flag to indicate that the parser is in the process of parsing a syntax
   * tree.
   */
  protected boolean setParsingFlag() {
    return this.isParsing.compareAndSet(false, true);
  }

  /**
   * Sets the 'parsing' flag to indicate that the parsing operation is NOT in progress.
   */
  protected boolean unsetParsingFlag() {
    return this.isParsing.compareAndSet(true, false);
  }

  /**
   * Request the parsing operation to be cancelled if the parser is in the process of parsing a
   * syntax tree.
   * <p>
   * The parse operation is NOT immediately cancelled.
   *
   * @return <code>true</code> if the cancellation was requested successfully, <code>false</code>
   * otherwise.
   */
  public boolean requestCancellation() {
    final var requested = Native.requestCancellation();
    setCancellationRequested(requested);
    return requested;
  }

  protected void setCancellationRequested(boolean isRequested) {
    cancellationLock.lock();
    try {
      this.isCancellationRequested.set(isRequested);
    } finally {
      cancellationLock.unlock();
    }
  }

  protected synchronized boolean isCancellationRequested() {
    cancellationLock.lock();
    try {
      return this.isCancellationRequested.get();
    } finally {
      cancellationLock.unlock();
    }
  }

  /**
   * Set the ranges of text that the parser should include when parsing.
   *
   * <p>By default, the parser will always include entire documents. This function allows you to
   * parse only a *portion* of a document but still return a syntax tree whose ranges match up with
   * the document as a whole. You can also pass multiple disjoint ranges.
   *
   * <p>If the ranges parameter is an empty array, then the entire document will be parsed.
   * Otherwise, the given ranges must be ordered from earliest to latest in the document, and they
   * must not overlap. That is, the following must hold for all `i` < `length - 1`:
   * ranges[i].end_byte <= ranges[i + 1].start_byte where `length` is the length of the ranges
   * array.
   *
   * <p>If this requirement is not satisfied, the operation will fail, the ranges will not be
   * assigned, and this function will return `false`. On success, this function returns `true`
   */
  public boolean setIncludedRanges(TSRange[] ranges) {
    checkAccess();
    return Native.setIncludedRanges(getNativeObject(), ranges);
  }

  public TSRange[] getIncludedRanges() {
    checkAccess();
    return Native.getIncludedRanges(getNativeObject());
  }

  /**
   * Instruct the parser to start the next parse from the beginning.
   *
   * <p>If the parser previously failed because of a timeout or a cancellation, then by default, it
   * will resume where it left off on the next call to any of the parsing functions. If you don't
   * want to resume, and instead intend to use this parser to parse some other document, you must
   * call this function first.
   */
  public void reset() {
    checkAccess();
    Native.reset(getNativeObject());
  }

  @Override
  protected void closeNativeObj() {
    Native.delete(getNativeObject());
  }

  private void throwIfParseNotCancelled() {
    if (isParsing() && !isCancellationRequested()) {
      throw new ParseInProgressException(
        "Parser is already parsing another syntax tree! Cancel the previous parse before starting another.");
    }
  }

  /**
   * Base class the {@link TSParser} exceptions.
   */
  protected static class TSParserException extends RuntimeException {

    public TSParserException(String message) {
      super(message);
    }

    public TSParserException(String message, Throwable cause) {
      super(message, cause);
    }

    public TSParserException(Throwable cause) {
      super(cause);
    }
  }

  /**
   * Thrown when a parse is requested while another parse is already in progress.
   */
  protected static final class ParseInProgressException extends TSParserException {

    public ParseInProgressException(String message) {
      super(message);
    }
  }

  private static class Native {

    public static native long newParser();

    public static native void delete(long parser);

    public static native void setLanguage(long parser, long language);

    public static native long getLanguage(long parser);

    public static native void reset(long parser);

    public static native void setTimeout(long parser, long timeout);

    public static native long getTimeout(long parser);

    public static native boolean setIncludedRanges(long parser, TSRange[] ranges);

    public static native TSRange[] getIncludedRanges(long parser);

    public static native long parse(long parser, long treePointer, long strPointer);

    public static native boolean requestCancellation();
  }
}

File: android-tree-sitter/src/main/cpp/ts_parser.cc
/*
 *  This file is part of android-tree-sitter.
 *
 *  android-tree-sitter library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2.1 of the License, or (at your option) any later version.
 *
 *  android-tree-sitter library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *  along with android-tree-sitter.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <atomic>
#include <mutex>
#include <iostream>

#include "utf16str/UTF16String.h"
#include "utils/ts_obj_utils.h"
#include "utils/ts_exceptions.h"
#include "utils/ts_preconditions.h"
#include "ts__log.h"

#include "ts_parser.h"

/**
 * `TSParserInternal` stores the actual tree sitter parser instance along
 * with the cancellation flag and the cancellation flag mutex.
 */
class TSParserInternal {
 public:

  TSParserInternal() {
    cancellation_flag_mutex = new std::mutex();
    cancellation_flag = new std::atomic<size_t *>(nullptr);
    parser = ts_parser_new();
  }

  ~TSParserInternal() {
    delete cancellation_flag_mutex;
    delete cancellation_flag;
    ts_parser_delete(parser);

    cancellation_flag_mutex = nullptr;
    cancellation_flag = nullptr;
    parser = nullptr;
  }

  TSParser *getParser(JNIEnv *env) {
    if (check_destroyed(env)) {
      return nullptr;
    }

    return this->parser;
  }

  bool begin_round(JNIEnv *env) {
    auto flag = get_cancellation_flag(env);

    if (flag) {
      throw_illegal_state(env,
                          "Parser is already parsing another syntax tree! You must cancel the current parse first!");
      return false;
    }

    // allocate a new cancellation flag
    flag = (size_t *) malloc(sizeof(int));
    set_cancellation_flag(env, flag);

    // set the cancellation flag to '0' to indicate that the parser should continue parsing
    *flag = 0;
    ts_parser_set_cancellation_flag(getParser(env), flag);

    return true;
  }

  void end_round(JNIEnv *env) {

    size_t *flag = get_cancellation_flag(env);

    // release the cancellation flag
    free((size_t *) flag);
    set_cancellation_flag(env, nullptr);
    ts_parser_set_cancellation_flag(getParser(env), nullptr);
  }

  size_t *get_cancellation_flag(JNIEnv *env) {
    if (check_destroyed(env)) {
      return nullptr;
    }

    std::lock_guard<std::mutex> get_lock(*cancellation_flag_mutex);
    return cancellation_flag->load();
  }

  void set_cancellation_flag(JNIEnv *env, size_t *flag) {
    if (check_destroyed(env)) {
      return;
    }

    std::lock_guard<std::mutex> set_lock(*cancellation_flag_mutex);
    cancellation_flag->store(flag);
  }

 private:
  std::mutex *cancellation_flag_mutex;
  std::atomic<size_t *> *cancellation_flag;

  TSParser *parser;

  bool check_destroyed(JNIEnv *env) {
    if (cancellation_flag_mutex == nullptr || cancellation_flag == nullptr
        || parser == nullptr) {
      throw_illegal_state(env, "TSParserInternal has already been destroyed");
      return true;
    }

    return false;
  }
};

static jlong
TSParser_newParser(JNIEnv *env,
                   jclass self) {
  auto parser = new TSParserInternal;
  return (jlong) parser;
}

static void
TSParser_delete(JNIEnv *env,
                jclass self,
                jlong parser_ptr) {
  req_nnp(env, parser_ptr);

  auto parser = (TSParserInternal *) parser_ptr;
  delete parser;
}

static void
TSParser_setLanguage(JNIEnv *env,
                     jclass self,
                     jlong parser,
                     jlong language) {
  req_nnp(env, parser, "parser");
  req_nnp(env, language, "language");
  ts_parser_set_language(((TSParserInternal *) parser)->getParser(env),
                         (TSLanguage *) language);
}

static jlong
TSParser_getLanguage(JNIEnv *env,
                     jclass self,
                     jlong parser) {
  req_nnp(env, parser);
  return (jlong) ts_parser_language(((TSParserInternal *) parser)->getParser(env));
}

static void
TSParser_reset(JNIEnv *env,
               jclass self,
               jlong parser) {
  req_nnp(env, parser);
  TSParser *pParser = ((TSParserInternal *) (parser))->getParser(env);
  LOGD("ts_parser.cc", "Reset parser: %p, language: %p", pParser,
       ts_parser_language(pParser));
  ts_parser_reset(pParser);
}

static void
TSParser_setTimeout(JNIEnv *env,
                    jclass self,
                    jlong parser,
                    jlong macros) {
  req_nnp(env, parser);
  ts_parser_set_timeout_micros(((TSParserInternal *) parser)->getParser(env),
                               macros);
}

static jlong
TSParser_getTimeout(JNIEnv *env,
                    jclass self,
                    jlong parser) {
  req_nnp(env, parser);
  return (jlong) ts_parser_timeout_micros(((TSParserInternal *) parser)->getParser(
      env));
}

static jboolean
TSParser_setIncludedRanges(
    JNIEnv *env,
    jclass self,
    jlong parser,
    jobjectArray ranges) {
  req_nnp(env, parser);
  int count = env->GetArrayLength(ranges);
  TSRange tsRanges[count];
  for (int i = 0; i < count; i++) {
    jobject range = env->GetObjectArrayElement(ranges, i);
    std::string msg = std::string("ranges[") + std::to_string(i) + "]";
    req_nnp(env, range, msg);
    tsRanges[i] = _unmarshalRange(env, range);
  }

  const TSRange *r = tsRanges;
  return (jboolean) ts_parser_set_included_ranges(((TSParserInternal *) parser)->getParser(
      env), r, count);
}

static jobjectArray
TSParser_getIncludedRanges(
    JNIEnv *env,
    jclass self,
    jlong parser) {
  req_nnp(env, parser);
  jint count;
  const TSRange *ranges =
      ts_parser_included_ranges(((TSParserInternal *) parser)->getParser(env),
                                reinterpret_cast<uint32_t *>(&count));
  jobjectArray result = createRangeArr(env, count);
  req_nnp(env, result, "TSRange[] from factory");

  for (uint32_t i = 0; i < count; i++) {
    const TSRange *r = (ranges + i);
    env->SetObjectArrayElement(result, (jint) i, _marshalRange(env, *r));
  }
  return result;
}

static jlong TSParser_parse(JNIEnv *env,
                            jclass clazz,
                            jlong parser,
                            jlong tree_pointer,
                            jlong str_pointer) {
  req_nnp(env, parser);
  req_nnp(env, str_pointer, "string");
  auto *ts_parser_internal = (TSParserInternal *) parser;
  TSParser *ts_parser = ts_parser_internal->getParser(env);
  TSTree *old_tree = tree_pointer == 0 ? nullptr : (TSTree *) tree_pointer;
  auto *source = as_str(env, str_pointer);

  if (!ts_parser_internal->begin_round(env)) {
    return 0;
  }

  auto src_cstring = source->to_cstring();

  // start parsing
  // if the user cancels the parse while this method is being executed
  // then this will return nullptr
  auto tree = ts_parser_parse_string_encoding(ts_parser,
                                              old_tree,
                                              src_cstring,
                                              source->byte_length(),
                                              TSInputEncodingUTF16);

  ts_parser_internal->end_round(env);
  delete[] src_cstring;


  return (jlong) tree;
}

static jboolean
TSParser_requestCancellation(
    JNIEnv *env,
    jclass clazz,
    jlong parser) {

  auto *parserInternal = (TSParserInternal *) parser;
  auto flag = parserInternal->get_cancellation_flag(env);

  // no parse is in progress
  if (flag == nullptr) {
    LOGD("TSParser",
         "Cannot cancel parsing, no parse is in progress (cancellation flag is nullptr).");
    return false;
  }

  // set the cancellation flag to a non-zero value to indicate that the parse
  // operation has been cancelled
  *flag = 1;
  LOGD("TSParser", "Cancellation flag has been set");
  return true;
}

void TSParser_Native__SetJniMethods(JNINativeMethod *methods, int count) {
  SET_JNI_METHOD(methods, TSParser_Native_newParser, TSParser_newParser);
  SET_JNI_METHOD(methods, TSParser_Native_delete, TSParser_delete);
  SET_JNI_METHOD(methods, TSParser_Native_setLanguage, TSParser_setLanguage);
  SET_JNI_METHOD(methods, TSParser_Native_getLanguage, TSParser_getLanguage);
  SET_JNI_METHOD(methods, TSParser_Native_reset, TSParser_reset);
  SET_JNI_METHOD(methods, TSParser_Native_setTimeout, TSParser_setTimeout);
  SET_JNI_METHOD(methods, TSParser_Native_getTimeout, TSParser_getTimeout);
  SET_JNI_METHOD(methods, TSParser_Native_setIncludedRanges, TSParser_setIncludedRanges);
  SET_JNI_METHOD(methods, TSParser_Native_getIncludedRanges, TSParser_getIncludedRanges);
  SET_JNI_METHOD(methods, TSParser_Native_parse, TSParser_parse);
  SET_JNI_METHOD(methods, TSParser_Native_requestCancellation,
                 TSParser_requestCancellation);
}